ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py

```diff
@@ -0,0 +1,327 @@
+import calendar
+import json
+import logging
+import os
+import sys
+import time
+from datetime import datetime, timedelta
+from typing import Any, TypedDict
+from urllib.error import URLError
+from urllib.parse import quote_plus
+from urllib.request import Request, urlopen
+
+import boto3
+from botocore.config import Config
+
+REGION = os.environ.get("AWS_CURRENT_REGION", "us-east-1")
+BFD_ENVIRONMENT = os.environ.get("BFD_ENVIRONMENT", "")
+ACCESS_JSON_LOG_GROUP = os.environ.get("ACCESS_JSON_LOG_GROUP", "")
+MESSAGES_JSON_LOG_GROUP = os.environ.get("MESSAGES_JSON_LOG_GROUP", "")
+LOG_LOOKBACK_SECONDS = int(os.environ.get("LOG_LOOKBACK_SECONDS", "330"))
+SLACK_WEBHOOK = os.environ.get("SLACK_WEBHOOK", "")
+BOTO_CONFIG = Config(
+    region_name=REGION,
+    # Instructs boto3 to retry upto 10 times using an exponential backoff
+    retries={
+        "total_max_attempts": 10,
+        "mode": "adaptive",
+    },
+)
+
+LOG_INSIGHTS_NOTIF_QUERY = """
+fields
+mdc.http_access_response_status as status,
+mdc.http_access_request_clientSSL_DN as partner,
+mdc.http_access_request_operation as op
+| filter status == 500
+| stats count(status) by partner, op
+""".strip()
+LOG_INSIGHTS_ACCESS_JSON_ERRORS_QUERY = """
+fields @message,
+mdc.http_access_response_status as status
+| filter status == 500
+| sort @timestamp desc
+""".strip()
+LOG_INSIGHTS_MESSAGES_JSON_ERRORS_QUERY = """
+fields @message
+| filter ispresent('exception')
+| filter ispresent('mdc.http_access_request_uri')
+| filter level = 'ERROR'
+| sort @timestamp desc
+""".strip()
+LOG_INSIGHTS_MESSAGES_JSON_CROSS_REF_QUERY = """
+fields @message,
+`mdc.http_access_request_header_Bluebutton-Originalquerytimestamp` as orig_timestamp
+| filter orig_timestamp == 'REPLACE WITH MDC ORIGINAL TIMESTAMP OF MESSAGE FROM ACCESS.JSON'
+""".strip()
```

Comment:
Is this query useful to include as part of the query links in the Slack notification? Log Insights has [no function for joining fields](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL_QuerySyntax.html) from two log sources (say, `messages.json` and `access.json`), so crafting a query to do this "cross-referencing" automatically is impossible. I wanted to include a _somewhat_ convenient way for an engineer responding to an alert to be able to see exception details for a particular `500` and this half-solution was all I could come up with.

The problem here is that each Log Insights URL added to the Slack message alert balloons its size considerably, as the query needs to be embedded within the URL that's posted with the alert. Slack messages have a [limit of about 4000 characters](https://api.slack.com/docs/rate-limits#rtm-posting-messages) and with this additional Log Insights query link the message is getting close to the limit; for example, an alert with a single partner table with only a single row is already 3636 characters long. Removing the link to this particular "cross-referencing" query reduces that to _2782 characters_, a difference of 854 characters.

Does anyone feel strongly that including a link to this query (that will query the time period associated with the alert) is useful enough to keep?

Reply:
Yeah, I'm thinking this particular query is probably more suited for inclusion in a runbook (which I also was considering writing but this story has already taken long enough as it is). However, I _do_ think the other two queries are useful as I suspect an interested engineer will want to quickly know _why_ we're returning errors and what the requests that are causing them look like. With the current alert, they could click one of the links and immediately be able to start investigating what's going on. With just a link to a runbook, there's just a _little_ extra friction there that I feel could lead to a less rapid response.

At the bare minimum, I'm going to remove `LOG_INSIGHTS_MESSAGES_JSON_CROSS_REF_QUERY` and all usages of it, including the link in the Slack message. I would like to keep the links to the other two queries, and maybe in the future, when the corresponding runbook exists, we can add a third link to it as well.